learning-at-home/hivemind
# Faster beam search: part1/2

This PR has exploded into >500 SLOC, so we decided to split it into two parts:
This part: 
  * better caching & key reuse in DHTNode
  * fix rare bug in traverse_dht that caused ~0.1% change of not finding an existing key
  * update benchmarks & tests

Second part:
  * hivemind.dht.DHT.get_best_experts, hivemind.dht.DHT.batch_get_best_experts
     * the new beam search supports time_budget to balance between latency and accuracy (see part2))
  * remove async def beam_search from hivemind.client.RemoteMixtureOfExperts, replace with DHT.get_best_experts
  * more tests

TODO
* [x] add option for limited time_budget in beam search
* [x] add option to reuse concurrent traverse_dht runs for the same key
* [x] add option for automatic cache refresh
* [x] deduplicate concurrent get requests
* add tests for all new functionality
  * [x] new caching
  * [x] key reuse
  * [x] LocalStorage updates
* [x] verify no performance/accuracy drawdown in tests/benchmark_dht.py
* [x] update benchmark_dht.py


__traverse_dht bugfix:__ previously, if a worker chose a heap where all candidates are farther than upper bound, it would immediately terminate search. That's bad because there could have been other workers that would have found closer keys and continued search. Now, a search can only terminate if there are no suitable candidates AND no concurrent workers that can affect that heap. As a result, benchmark_dht now has exactly 100% accuracy for get operations at 1000+ nodes.

`python benchmark_dht.py --num_peers 1024 --num_experts 16384 --expert_batch_size 64 --expiration 99999 --increase_file_limit`
Before:
```
Increasing file limit - soft 1024=>32768, hard 1048576=>32768
Creating peers...
100%|███████████████████████████████████████████████████| 1024/1024 [01:51<00:00,  9.22it/s]
Sampled 16384 unique ids (after deduplication)
Storing peers to dht in batches of 64...
100%|█████████████████████████████████████████████████████| 256/256 [13:00<00:00,  3.05s/it]
Store success rate: 100.0% (48904 / 48904)
Mean store time: 0.015967, Total: 780.85
100%|█████████████████████████████████████████████████████| 256/256 [02:01<00:00,  2.11it/s]
Get success rate: 100.0 (16380 / 16384)
Mean get time: 0.00740, Total: 121.29011
Node survival rate: 100.000%
```

After:
```
Increasing file limit - soft 1024=>32768, hard 1048576=>32768
Creating peers...
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1024/1024 [01:45<00:00,  9.74it/s]
Sampled 16384 unique ids (after deduplication)
Storing peers to dht in batches of 64...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 256/256 [12:07<00:00,  2.84s/it]
Store success rate: 100.0% (48920 / 48920)
Mean store time: 0.01487, Total: 727.46
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 256/256 [01:48<00:00,  2.35it/s]
Get success rate: 100.0 (16384 / 16384)
Mean get time: 0.00664, Total: 108.73952
Node survival rate: 100.000%
```


__Automatic cache refresh:__ every time DHTNode.get_* hits a local node's cache, it will mark this key as active. If an active key approaches expiration, DHTNode will launch a background task to fetch updates to that key. As a result, hivemind.DHT.get_experts and hivemind.DHT.first_k_active will no longer have temporary spikes in latency when cache gets invalidated.

__Key reuse:__ Old behavior: if you request the same key from DHT multiple times before the first request finishes, all requests would trigger a full DHT lookup and not benefit from caching (since cache wasn't available yet).
New behavior: DHTNode.get_* will check if there already are pending traverse_dht requests for the same key. If so, when one of these pending requests finishes, it will also return result to other requests for that key as long as expiration time is compatible.
